src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -94,16 +141,38 @@
                 refEx = answer.refEx;
             }
 
+            this.enumCtx = new EnumCtx(homeCtx, answer, homeCtx.clnt);
             // Ensures that context won't get closed from underneath us
-            this.homeCtx = homeCtx;
-            homeCtx.incEnumCount();
-            enumClnt = homeCtx.clnt; // remember
+            this.enumCtx.homeCtx.incEnumCount();
+            this.cleanable = CleanerFactory.cleaner().register(this, this.enumCtx);
+            // For finalizers, JLS guarantees that the constructor of a finalizeable
+            // class will complete before the finalizer is run. Classes that use
+            // Cleaner instead don't currently have such a guarantee.
+
+            // 'this' needs to stay reachable until registration with Cleaner
+            // completes, to ensure that the cleanable state that has been created
+            // will eventually be cleaned up. Within Cleaner.register(), there
+            // are reachability fences to ensure that the registered object remains
+            // reachable.
+            // TODO: Is anything else needed so that this constructor
+            //  "happens-before" the cleaning action ?
     }
 
     @Override
     public final T nextElement() {
         try {
-            return next();
+            try {
+                return next();
+            } finally {
+                // See comment in nextImpl(). This is similar, but in this case, next()
+                // *is* overridable. Fences are included here, in case next() is
+                // overridden to access the cleanable state without the proper fences.
+
+                // Ensure writes are visible to the Cleaner thread
+                VarHandle.fullFence();
```

Comment:
I don't think any memory fences are needed here. Reachability fence is enough. Cleaner runs in its own thread by polling enqued PhantomReference(s) from the ReferenceQueue. There is a happens-before edge between ReferenceHandler thread enqueue-ing "pending" PhantomReferences and Cleaner thread dequeue-ing them. There is a happens-before edge between GC thread clearing a PhantomReference and linking it into a "pending" list and ReferenceHandler thread unlinking it from the "pending" list and enque-ing it into a ReferenceQueue. There is a happens-before edge before a call to Reference.reachabilityFence(referent) and a GC thread discovering a phantom-reachable referent and clearing the PhantomReference and linking it into a "pending" list.
So there you have a happens-before chain which makes all writes before a call to eference.reachabilityFence(this) visible to a Cleaner task that is initialized to observe reachabillity of this....